File: PixelSpriteEditor/src/EraserTool.java
/**
 * Same drag/interpolation behavior as PencilTool, but always writes fully
 * transparent (ARGB 0) instead of the current color.
 */
public class EraserTool implements Tool {
    private static final int TRANSPARENT = 0;

    private int lastX, lastY;
    private boolean hasLast;

    @Override
    public void onPress(EditorState state, int x, int y) {
        state.beginStroke();
        state.paintPixel(x, y, TRANSPARENT);
        lastX = x;
        lastY = y;
        hasLast = true;
    }

    @Override
    public void onDrag(EditorState state, int x, int y) {
        if (!hasLast) {
            lastX = x;
            lastY = y;
            hasLast = true;
        }
        LineUtil.forEachLinePixel(lastX, lastY, x, y, (px, py) -> state.paintPixel(px, py, TRANSPARENT));
        lastX = x;
        lastY = y;
    }

    @Override
    public void onRelease(EditorState state, int x, int y) {
        state.endStroke();
        hasLast = false;
    }
}

File: legacy-cpp-sdl2/src/Tools/EraserTool.cpp
#include "EraserTool.h"

#include "App.h"
#include "LineUtil.h"

namespace {
constexpr Color kTransparent{0, 0, 0, 0};
}

void EraserTool::OnMouseDown(App& app, int x, int y) {
    app.BeginStroke();
    app.PaintPixel(x, y, kTransparent);
    lastX_ = x;
    lastY_ = y;
    hasLast_ = true;
}

void EraserTool::OnMouseDrag(App& app, int x, int y) {
    if (!hasLast_) {
        lastX_ = x;
        lastY_ = y;
        hasLast_ = true;
    }
    ForEachLinePixel(lastX_, lastY_, x, y, [&](int px, int py) { app.PaintPixel(px, py, kTransparent); });
    lastX_ = x;
    lastY_ = y;
}

void EraserTool::OnMouseUp(App& app, int /*x*/, int /*y*/) {
    app.EndStroke();
    hasLast_ = false;
}
